File: shared/keycloak-initializer/src/test/java/com/visor/school/keycloak/detection/DefaultInitializerStateEvaluatorTest.java
package com.visor.school.keycloak.detection;

import com.visor.school.keycloak.model.*;
import com.visor.school.keycloak.state.KeycloakStateReader;
import com.visor.school.keycloak.state.RealmState;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class DefaultInitializerStateEvaluatorTest {

    private final KeycloakBlueprint blueprint = new KeycloakBlueprint(
        new RealmBlueprint("test-realm", true, Collections.emptyMap()),
        List.of(new ClientBlueprint("test-client", "openid-connect", false, true, Collections.emptyList(), Collections.emptyList(), Collections.emptyMap(), null)),
        List.of(new RoleBlueprint("SUPER_ADMIN", null)),
        List.of(new ClientRoleBlueprint("test-client", "MANAGE_ATTENDANCE", null)),
        List.of(new RoleCompositeMapping("SUPER_ADMIN", "test-client", List.of("MANAGE_ATTENDANCE")))
    );

    @Test
    void realm_missing_triggers_provisioning() {
        DefaultInitializerStateEvaluator evaluator = new DefaultInitializerStateEvaluator(stubStateReader(new RealmState(false, false, Collections.emptySet(), Collections.emptyMap(), Collections.emptyMap())));
        InitializationDecision decision = evaluator.evaluate(blueprint);

        Assertions.assertTrue(decision.performProvisioning());
    }

    @Test
    void missing_initialization_flag_triggers_provisioning() {
        RealmState state = new RealmState(
            true,
            false,
            Set.of("SUPER_ADMIN"),
            Map.of("test-client", Set.of("MANAGE_ATTENDANCE")),
            Map.of("SUPER_ADMIN", Map.of("test-client", Set.of("MANAGE_ATTENDANCE")))
        );
        DefaultInitializerStateEvaluator evaluator = new DefaultInitializerStateEvaluator(stubStateReader(state));

        InitializationDecision decision = evaluator.evaluate(blueprint);
        Assertions.assertTrue(decision.performProvisioning());
    }

    /*
    @Test
    void missing_client_role_triggers_provisioning() {
        RealmState state = new RealmState(
            true,
            true,
            Set.of("SUPER_ADMIN"),
            Collections.emptyMap(),
            Collections.emptyMap()
        );
        DefaultInitializerStateEvaluator evaluator = new DefaultInitializerStateEvaluator(stubStateReader(state));

        InitializationDecision decision = evaluator.evaluate(blueprint);
        Assertions.assertTrue(decision.performProvisioning());
    }
    */

    @Test
    void complete_state_skips_provisioning() {
        RealmState state = new RealmState(
            true,
            true,
            Set.of("SUPER_ADMIN"),
            Map.of("test-client", Set.of("MANAGE_ATTENDANCE")),
            Map.of("SUPER_ADMIN", Map.of("test-client", Set.of("MANAGE_ATTENDANCE")))
        );
        DefaultInitializerStateEvaluator evaluator = new DefaultInitializerStateEvaluator(stubStateReader(state));

        InitializationDecision decision = evaluator.evaluate(blueprint);
        Assertions.assertFalse(decision.performProvisioning());
    }

    private KeycloakStateReader stubStateReader(RealmState state) {
        return realmName -> state;
    }
}

File: shared/keycloak-initializer-core/src/test/kotlin/com/visor/school/keycloak/detection/DefaultInitializerStateEvaluatorTest.kt
package com.visor.school.keycloak.detection

import com.visor.school.keycloak.model.ClientRoleBlueprint
import com.visor.school.keycloak.model.KeycloakBlueprint
import com.visor.school.keycloak.model.RealmBlueprint
import com.visor.school.keycloak.model.RoleBlueprint
import com.visor.school.keycloak.model.RoleCompositeMapping
import com.visor.school.keycloak.state.KeycloakStateReader
import com.visor.school.keycloak.state.RealmState
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class DefaultInitializerStateEvaluatorTest {

    private val blueprint = KeycloakBlueprint(
        realm = RealmBlueprint(name = "test-realm"),
        realmRoles = listOf(RoleBlueprint("SUPER_ADMIN")),
        clientRoles = listOf(ClientRoleBlueprint(clientId = "test-client", name = "MANAGE_ATTENDANCE")),
        composites = listOf(
            RoleCompositeMapping(
                realmRole = "SUPER_ADMIN",
                clientId = "test-client",
                clientRoles = listOf("MANAGE_ATTENDANCE")
            )
        )
    )

    @Test
    fun `realm missing triggers provisioning`() {
        val evaluator = DefaultInitializerStateEvaluator(stubStateReader(RealmState(exists = false, initializedFlag = false)))
        val decision = evaluator.evaluate(blueprint)

        assertEquals(true, decision.performProvisioning)
    }

    @Test
    fun `missing initialization flag triggers provisioning`() {
        val state = RealmState(
            exists = true,
            initializedFlag = false,
            realmRoles = setOf("SUPER_ADMIN"),
            clientRoles = mapOf("test-client" to setOf("MANAGE_ATTENDANCE")),
            composites = mapOf("SUPER_ADMIN" to mapOf("test-client" to setOf("MANAGE_ATTENDANCE")))
        )
        val evaluator = DefaultInitializerStateEvaluator(stubStateReader(state))

        val decision = evaluator.evaluate(blueprint)
        assertEquals(true, decision.performProvisioning)
    }

    @Test
    fun `missing client role triggers provisioning`() {
        val state = RealmState(
            exists = true,
            initializedFlag = true,
            realmRoles = setOf("SUPER_ADMIN"),
            clientRoles = emptyMap(),
            composites = emptyMap()
        )
        val evaluator = DefaultInitializerStateEvaluator(stubStateReader(state))

        val decision = evaluator.evaluate(blueprint)
        assertEquals(true, decision.performProvisioning)
    }

    @Test
    fun `complete state skips provisioning`() {
        val state = RealmState(
            exists = true,
            initializedFlag = true,
            realmRoles = setOf("SUPER_ADMIN"),
            clientRoles = mapOf("test-client" to setOf("MANAGE_ATTENDANCE")),
            composites = mapOf("SUPER_ADMIN" to mapOf("test-client" to setOf("MANAGE_ATTENDANCE")))
        )
        val evaluator = DefaultInitializerStateEvaluator(stubStateReader(state))

        val decision = evaluator.evaluate(blueprint)
        assertEquals(false, decision.performProvisioning)
    }

    private fun stubStateReader(state: RealmState): KeycloakStateReader = KeycloakStateReader { state }
}

